No support for move-only types

te doesn't seem to support move-only types, even though in the presence of move constructor and rvalue it will move the object and not use the copy constructor.

For example, given the following code that prints all calls to member functions of a tested class Circle:
```
#include "external/te/include/te.hpp"
#include <iostream>

using std::cout;

struct Drawable
{
    void draw(std::ostream &out) const
    {
        te::call([](auto const &self, auto &out) { self.draw(out); }, *this, out);
    }
};

struct Circle
{
    Circle() { cout << "Circle()\n"; }
    Circle(const Circle&) { cout << "Circle(const Circle&) ctor\n"; }
    Circle& operator=(const Circle&) { cout << "operator=(const Circle&)\n"; }
    Circle(Circle&&) { cout << "Circle(Circle&&)\n"; }
    Circle& operator=(Circle&&) { cout << "operator=(Circle&&)\n"; }
    ~Circle() { cout << "~Circle()\n"; }

    void draw(std::ostream &out) const { out << "Circle::draw\n"; }
};

void draw(const te::poly<Drawable>& drawable) { drawable.draw(cout); }

int main()
{
    Circle circle{};
    draw(std::move(circle));
}
````
we will get this output:
```
Circle()
Circle(Circle&&)
Circle::draw
~Circle()
~Circle()
```

But when we try to mark the copy constructor and copy assignment operator as deleted as below:
```
struct Circle
{
    Circle() { cout << "Circle()\n"; }
    Circle(const Circle&) = delete;
    Circle& operator=(const Circle&) = delete;
    Circle(Circle&&) { cout << "Circle(Circle&&)\n"; }
    Circle& operator=(Circle&&) { cout << "operator=(Circle&&)\n"; }
    ~Circle() { cout << "~Circle()\n"; }

    void draw(std::ostream &out) const { out << "Circle::draw\n"; }
};
```
we will get a compilation error: 
```
main.cpp:33:19: error: invalid initialization of reference of type ‘const te::v1::poly<Drawable>&’ from expression of type ‘std::remove_reference<Circle&>::type {aka Circle}’
     draw(std::move(circle));
          ~~~~~~~~~^~~~~~~~
main.cpp:28:6: note: in passing argument 1 of ‘void draw(const te::v1::poly<Drawable>&)’
 void draw(const te::poly<Drawable>& drawable) { drawable.draw(cout); }
      ^~~~
```
Are there any plans to support noncopyable types?

**Environment**
OS: Ubuntu 17.10 x64
Compiler: GCC 7.2 x64

--- a/include/te.hpp
+++ b/include/te.hpp
@@ -146,10 +146,11 @@ class dynamic_storage {
  public:
   template <class T, class T_ = std::decay_t<T> >
   constexpr explicit dynamic_storage(T &&t, detail::void_ptr &ptr) noexcept {
-    ptr.reset(
-        new T_{std::forward<T>(t)},
-        [](void *ptr) { delete static_cast<T_ *>(ptr); },
-        [](void *ptr) -> void * { return new T_{*static_cast<T_ *>(ptr)}; });
+    ptr.reset(new T_{std::forward<T>(t)},
+              [](void *ptr) { delete static_cast<T_ *>(ptr); },
+              [](void *ptr) -> void * {
+                return new T_{std::move(*static_cast<T_ *>(ptr))};
+              });
   }
 };
 
@@ -162,7 +163,7 @@ class local_storage {
     new (&data) T_{std::forward<T>(t)};
     ptr.reset(&data, [](void *ptr) { static_cast<T_ *>(ptr)->~T_(); },
               [](void *ptr) -> void * {
-                return new (ptr) T_{*static_cast<T_ *>(ptr)};
+                return new (ptr) T_{std::move(*static_cast<T_ *>(ptr))};
               });
   }
 
@@ -197,8 +198,9 @@ class poly : detail::poly_base,
  public:
   template <class T,
             class = std::enable_if_t<not std::is_convertible<T, poly>{} and
-                                     std::is_copy_constructible<T>{} and
-                                     std::is_destructible<T>{}> >
+                                     std::is_destructible<T>{} and
+                                     (std::is_copy_constructible<T>{} or
+                                      std::is_move_constructible<T>{})> >
   constexpr poly(T &&t) noexcept
       : poly{std::forward<T>(t),
              detail::type_list<decltype(detail::requires__<I>(bool{}))>{}} {}

--- a/test/te.cpp
+++ b/test/te.cpp
@@ -154,6 +154,62 @@ test should_reassign = [] {
   }
 };
 
+struct CircleNoncopyable {
+  CircleNoncopyable() = default;
+  CircleNoncopyable(const CircleNoncopyable &) = delete;
+  CircleNoncopyable &operator=(const CircleNoncopyable &) = delete;
+  CircleNoncopyable(CircleNoncopyable &&) = default;
+  CircleNoncopyable &operator=(CircleNoncopyable &&) = default;
+  ~CircleNoncopyable() noexcept = default;
+
+  void draw(std::ostream &out) const { out << "Circle"; }
+};
+
+struct SquareNoncopyable {
+  SquareNoncopyable() = default;
+  SquareNoncopyable(const SquareNoncopyable &) = delete;
+  SquareNoncopyable &operator=(const SquareNoncopyable &) = delete;
+  SquareNoncopyable(SquareNoncopyable &&) = default;
+  SquareNoncopyable &operator=(SquareNoncopyable &&) = default;
+  ~SquareNoncopyable() noexcept = default;
+
+  void draw(std::ostream &out) const { out << "Square"; }
+};
+
+test should_support_movable_only_types = [] {
+  te::poly<Drawable> drawable = CircleNoncopyable{};
+
+  {
+    std::stringstream str{};
+    drawable.draw(str);
+    expect("Circle" == str.str());
+  }
+
+  {
+    std::stringstream str{};
+    drawable = SquareNoncopyable{};
+    drawable.draw(str);
+    expect("Square" == str.str());
+  }
+};
+
+test should_support_movable_only_types_with_local_storage = [] {
+  te::poly<Drawable, te::local_storage<32>> drawable = CircleNoncopyable{};
+
+  {
+    std::stringstream str{};
+    drawable.draw(str);
+    expect("Circle" == str.str());
+  }
+
+  {
+    std::stringstream str{};
+    drawable = SquareNoncopyable{};
+    drawable.draw(str);
+    expect("Square" == str.str());
+  }
+};
+
 test should_support_containers = [] {
   std::vector<te::poly<Drawable>> drawables{};
 

